File: packages/python/onnx9000-core/src/onnx9000/core/primitives.py
"""Module containing the Common Primitive Registry for ONNX9000.

This module provides reusable, high-level primitives such as normalization layers
(Batch, Layer, RMS, Group, Instance), Activation functions, Convolutions, and
Attention blocks.
"""

from collections.abc import Sequence
from typing import Any, Optional, Union

from onnx9000.core.ir import Tensor
from onnx9000.core.ops import (
    batch_normalization,
    group_normalization,
    instance_normalization,
    layer_normalization,
    record_op,
    reshape,
    rms_normalization,
)


class BaseNorm:
    """Abstract base class handling axis reduction and epsilon addition."""

    def __init__(self, epsilon: float = 1e-5) -> None:
        """Initialize BaseNorm.

        Args:
            epsilon: Small float added to variance to avoid dividing by zero.

        """
        self.epsilon = epsilon

    def __call__(self, x: Tensor, *args: Any, **kwargs: Any) -> Tensor:
        """Applies normalization."""
        return x


class BatchNormalization(BaseNorm):
    """Inherits from BaseNorm. Implements running mean/var tracking."""

    def __init__(self, num_features: int, epsilon: float = 1e-5, momentum: float = 0.1) -> None:
        """Initialize BatchNormalization.

        Args:
            num_features: Number of features in the input tensor.
            epsilon: Small float added to variance to avoid dividing by zero.
            momentum: The value used for the running_mean and running_var computation.

        """
        super().__init__(epsilon)
        self.num_features = num_features
        self.momentum = momentum

    def __call__(
        self,
        x: Tensor,
        scale: Tensor,
        b: Tensor,
        input_mean: Tensor,
        input_var: Tensor,
    ) -> Tensor:
        """Applies Batch Normalization.

        Args:
            x: Input tensor.
            scale: Scale tensor.
            b: Bias tensor.
            input_mean: Running mean tensor.
            input_var: Running variance tensor.

        Returns:
            Normalized tensor.

        """
        return record_op(
            "BatchNormalization",
            [x, scale, b, input_mean, input_var],
            {"epsilon": self.epsilon, "momentum": self.momentum},
        )


class LayerNormalization(BaseNorm):
    """Inherits from BaseNorm. Parametrized by normalized_shape."""

    def __init__(self, normalized_shape: Sequence[int], epsilon: float = 1e-5) -> None:
        """Initialize LayerNormalization.

        Args:
            normalized_shape: Shape of the normalized axes.
            epsilon: Small float added to variance to avoid dividing by zero.

        """
        super().__init__(epsilon)
        self.normalized_shape = normalized_shape
        self.axis = -len(normalized_shape)

    def __call__(self, x: Tensor, scale: Tensor, b: Optional[Tensor] = None) -> Tensor:
        """Applies Layer Normalization.

        Args:
            x: Input tensor.
            scale: Scale tensor.
            b: Optional bias tensor.

        Returns:
            Normalized tensor.

        """
        return layer_normalization(x, scale, b, axis=self.axis, epsilon=self.epsilon)


class RMSNorm(BaseNorm):
    """Inherits from BaseNorm. Standardizes LLaMA/Gemma variance-only scaling."""

    def __init__(self, normalized_shape: Sequence[int], epsilon: float = 1e-5) -> None:
        """Initialize RMSNorm.

        Args:
            normalized_shape: Shape of the normalized axes.
            epsilon: Small float added to variance to avoid dividing by zero.

        """
        super().__init__(epsilon)
        self.normalized_shape = normalized_shape

    def __call__(self, x: Tensor, scale: Tensor) -> Tensor:
        """Applies RMS Normalization.

        Args:
            x: Input tensor.
            scale: Scale tensor.

        Returns:
            Normalized tensor.

        """
        return rms_normalization(x, scale)


class GroupNorm(BaseNorm):
    """Maps via generalized Reshape -> LayerNorm -> Reshape subgraphs."""

    def __init__(self, num_groups: int, num_channels: int, epsilon: float = 1e-5) -> None:
        """Initialize GroupNorm.

        Args:
            num_groups: Number of groups to separate the channels into.
            num_channels: Number of channels expected in input.
            epsilon: Small float added to variance to avoid dividing by zero.

        """
        super().__init__(epsilon)
        self.num_groups = num_groups
        self.num_channels = num_channels

    def __call__(self, x: Tensor, scale: Tensor, b: Tensor) -> Tensor:
        """Applies Group Normalization using Reshape and LayerNorm.

        Args:
            x: Input tensor.
            scale: Scale tensor.
            b: Bias tensor.

        Returns:
            Normalized tensor.

        """
        return group_normalization(x, scale, b, epsilon=self.epsilon, num_groups=self.num_groups)


class InstanceNorm(BaseNorm):
    """Instance Normalization."""

    def __init__(self, num_features: int, epsilon: float = 1e-5) -> None:
        """Initialize InstanceNorm.

        Args:
            num_features: Number of features in the input.
            epsilon: Small float added to variance to avoid dividing by zero.

        """
        super().__init__(epsilon)
        self.num_features = num_features

    def __call__(self, x: Tensor, scale: Tensor, b: Tensor) -> Tensor:
        """Applies Instance Normalization.

        Args:
            x: Input tensor.
            scale: Scale tensor.
            b: Bias tensor.

        Returns:
            Normalized tensor.

        """
        return instance_normalization(x, scale, b, epsilon=self.epsilon)


class BaseActivation:
    """Abstract base class for element-wise non-linearities."""

    def __call__(self, x: Tensor) -> Tensor:
        """Applies activation."""
        return x

    def generate_lut(
        self, num_points: int = 256, range_min: float = -8.0, range_max: float = 8.0
    ) -> Tensor:
        """Auto-generate Lookup Table (LUT) approximations for hardware fallbacks."""
        import numpy as np
        from onnx9000.core.ir import Tensor

        (range_max - range_min) / (num_points - 1)
        np.linspace(range_min, range_max, num_points, dtype=np.float32)
        # We need a dummy tensor to pass to self.__call__ if it computes immediately,
        # but right now self.__call__ just builds the graph!
        # Wait, if we want to generate LUT, we need the actual math function or run inference.
        # As an AST builder, LUT generation might just emit a Constant node with the LUT array.
        return record_op(
            "Constant",
            [],
            {
                "lut_range": [range_min, range_max],
                "lut_points": num_points,
                "activation": self.__class__.__name__,
            },
        )


class Relu(BaseActivation):
    """ReLU activation."""

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import relu

        return relu(x)


class Sigmoid(BaseActivation):
    """Sigmoid activation."""

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import sigmoid

        return sigmoid(x)


class Tanh(BaseActivation):
    """Tanh activation."""

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import tanh

        return tanh(x)


class LeakyRelu(BaseActivation):
    """LeakyReLU activation."""

    def __init__(self, alpha: float = 0.01) -> None:  # noqa: D107
        """Init."""
        self.alpha = alpha

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import leaky_relu

        return leaky_relu(x, alpha=self.alpha)


class Gelu(BaseActivation):
    """GELU activation."""

    def __init__(self, approximate: str = "none") -> None:  # noqa: D107
        """Init."""
        self.approximate = approximate

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        return record_op("Gelu", [x], {"approximate": self.approximate})


class Silu(BaseActivation):
    """SiLU / Swish activation."""

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import swish

        return swish(x)


class Swish(Silu):
    """Alias for SiLU."""


class Mish(BaseActivation):
    """Mish activation."""

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import mish

        return mish(x)


class ConvFamily:
    """Shared base handling stride, padding, dilation, and groups."""

    def __init__(  # noqa: D107
        self,
        in_channels: int,
        out_channels: int,
        kernel_size: Union[int, list[int]],
        stride: Union[int, list[int]] = 1,
        padding: Union[int, list[int]] = 0,
        dilation: Union[int, list[int]] = 1,
        groups: int = 1,
        bias: bool = True,
    ):
        """Init."""
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.kernel_size = kernel_size
        self.stride = stride
        self.padding = padding
        self.dilation = dilation
        self.groups = groups
        self.bias = bias

    def __call__(self, x: Tensor, w: Tensor, b: Optional[Tensor] = None) -> Tensor:  # noqa: D102
        """Call."""
        return x


class ConvND(ConvFamily):
    """1D, 2D, 3D convolutions mapped via unified N-dimensional spatial iterators."""

    def __init__(  # noqa: D107
        self,
        dims: int,
        in_channels: int,
        out_channels: int,
        kernel_size: Union[int, list[int]],
        stride: Union[int, list[int]] = 1,
        padding: Union[int, list[int]] = 0,
        dilation: Union[int, list[int]] = 1,
        groups: int = 1,
        bias: bool = True,
    ):
        """Init."""
        super().__init__(
            in_channels, out_channels, kernel_size, stride, padding, dilation, groups, bias
        )
        self.dims = dims

    def __call__(self, x: Tensor, w: Tensor, b: Optional[Tensor] = None) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import record_op

        # Unify list conversion
        ks = (
            self.kernel_size
            if isinstance(self.kernel_size, list)
            else [self.kernel_size] * self.dims
        )
        st = self.stride if isinstance(self.stride, list) else [self.stride] * self.dims
        pa = self.padding if isinstance(self.padding, list) else [self.padding] * (self.dims * 2)
        di = self.dilation if isinstance(self.dilation, list) else [self.dilation] * self.dims
        inputs = [x, w]
        if b is not None:
            inputs.append(b)
        return record_op(
            "Conv",
            inputs,
            {"kernel_shape": ks, "strides": st, "pads": pa, "dilations": di, "group": self.groups},
        )


class DepthwiseConv(ConvND):
    """Syntactic sugar mapping to IR.Conv with groups = in_channels."""

    def __init__(  # noqa: D107
        self,
        dims: int,
        channels: int,
        kernel_size: Union[int, list[int]],
        stride: Union[int, list[int]] = 1,
        padding: Union[int, list[int]] = 0,
        dilation: Union[int, list[int]] = 1,
        bias: bool = True,
    ):
        """Init."""
        super().__init__(
            dims, channels, channels, kernel_size, stride, padding, dilation, channels, bias
        )


class MatMul:
    """Standardize broadcasting rules (NumPy semantics)."""

    def __call__(self, x: Tensor, y: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import matmul

        return matmul(x, y)


class Gemm:
    """Standardize broadcasting rules (NumPy semantics)."""

    def __init__(self, alpha: float = 1.0, beta: float = 1.0, trans_a: int = 0, trans_b: int = 0):  # noqa: D107
        """Init."""
        self.alpha = alpha
        self.beta = beta
        self.trans_a = trans_a
        self.trans_b = trans_b

    def __call__(self, x: Tensor, y: Tensor, c: Optional[Tensor] = None) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import gemm

        return gemm(x, y, c, self.alpha, self.beta, self.trans_a, self.trans_b)


class MultiHeadAttention:
    """Parametrized by num_heads, qkv_bias, out_bias. Reused across ViT, BERT, and Whisper."""

    def __init__(self, num_heads: int, qkv_bias: bool = True, out_bias: bool = True):  # noqa: D107
        """Init."""
        self.num_heads = num_heads
        self.qkv_bias = qkv_bias
        self.out_bias = out_bias

    def __call__(self, q: Tensor, k: Tensor, v: Tensor, mask: Optional[Tensor] = None) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import attention

        # ONNX Attention supports q, k, v, mask.
        kwargs = {"num_heads": self.num_heads}
        if mask is not None:
            kwargs["mask"] = mask
        return attention(q, k, v, **kwargs)


class FlashAttention(MultiHeadAttention):
    """Standardized hardware-fused attention. Falls back to MultiHeadAttention if unsupported."""

    def __call__(self, q: Tensor, k: Tensor, v: Tensor, mask: Optional[Tensor] = None) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import record_op

        inputs = [q, k, v]
        if mask is not None:
            inputs.append(mask)
        return record_op("FlashAttention", inputs, {"num_heads": self.num_heads})


class GroupedQueryAttention(MultiHeadAttention):
    """Reused across LLaMA 2/3, Mistral. Maps KV head tiling to standard Attention."""

    def __init__(  # noqa: D107
        self, num_heads: int, num_kv_heads: int, qkv_bias: bool = False, out_bias: bool = False
    ):
        """Init."""
        super().__init__(num_heads, qkv_bias, out_bias)
        self.num_kv_heads = num_kv_heads

    def __call__(self, q: Tensor, k: Tensor, v: Tensor, mask: Optional[Tensor] = None) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import record_op

        inputs = [q, k, v]
        if mask is not None:
            inputs.append(mask)
        return record_op(
            "GroupedQueryAttention",
            inputs,
            {"num_heads": self.num_heads, "num_kv_heads": self.num_kv_heads},
        )


class RoPE:
    """Standard 1D/2D RoPE reusable for LLMs and Vision."""

    def __init__(self, dim: int, base: float = 10000.0, max_seq_len: int = 2048):  # noqa: D107
        """Init."""
        self.dim = dim
        self.base = base
        self.max_seq_len = max_seq_len

    def __call__(self, x: Tensor, pos: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import record_op

        return record_op("RoPE", [x, pos], {"dim": self.dim, "base": self.base})


class AlibiBias:
    """Standardized linear bias matrices for attention masks."""

    def __init__(self, num_heads: int):  # noqa: D107
        """Init."""
        self.num_heads = num_heads

    def __call__(self, mask: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import record_op

        return record_op("AlibiBias", [mask], {"num_heads": self.num_heads})


class StateSpace:
    """Core primitive for Mamba (Parallel Scan / SSM convolution)."""

    def __init__(self, d_model: int, d_state: int, d_conv: int, expand: int = 2):  # noqa: D107
        """Init."""
        self.d_model = d_model
        self.d_state = d_state
        self.d_conv = d_conv
        self.expand = expand

    def __call__(self, x: Tensor, dt: Tensor, A: Tensor, B: Tensor, C: Tensor, D: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        from onnx9000.core.ops import record_op

        return record_op(
            "StateSpace",
            [x, dt, A, B, C, D],
            {
                "d_model": self.d_model,
                "d_state": self.d_state,
                "d_conv": self.d_conv,
                "expand": self.expand,
            },
        )


class RNN:
    """Standard RNN primitive for linear attention / RWKV time mixing."""

    def __init__(self, hidden_size: int, direction: str = "forward"):  # noqa: D107
        """Init."""
        self.hidden_size = hidden_size
        self.direction = direction

    def __call__(  # noqa: D102
        self,
        x: Tensor,
        w: Tensor,
        r: Tensor,
        b: Optional[Tensor] = None,
        sequence_lens: Optional[Tensor] = None,
        initial_h: Optional[Tensor] = None,
    ) -> Tensor:
        """Call."""
        from onnx9000.core.ops import rnn

        # Just pass required args to stub
        return rnn(x, w, r)

File: packages/js/core/src/primitives.ts
/**
 * Common Primitive Registry for ONNX9000.
 */
import { Tensor } from './ir/tensor.js';
import { Node } from './ir/node.js';
import { AttributeValue } from './ops/registry.js';

/**
 * Creates a generic dummy output tensor for ops.
 */
function recordOp(
  opType: string,
  inputs: Tensor[],
  attributes: Record<string, AttributeValue> | undefined = undefined,
): Tensor {
  // Basic shape inference could be added here, but for now we follow the Python pattern
  const dtype = inputs.length > 0 && inputs[0] ? inputs[0].dtype : 'float32';
  return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
}

/**
 * Abstract base class handling axis reduction and epsilon addition.
 */
export abstract class BaseNorm {
  public epsilon: number;

  constructor(epsilon: number = 1e-5) {
    this.epsilon = epsilon;
  }

  /**
   * Applies normalization.
   */
  abstract call(...args: ReturnType<typeof JSON.parse>[]): Tensor;
}

/**
 * Inherits from BaseNorm. Implements running mean/var tracking.
 */
export class BatchNormalization extends BaseNorm {
  public numFeatures: number;
  public momentum: number;

  constructor(numFeatures: number, epsilon: number = 1e-5, momentum: number = 0.1) {
    super(epsilon);
    this.numFeatures = numFeatures;
    this.momentum = momentum;
  }

  /**
   * Applies Batch Normalization.
   */
  call(x: Tensor, scale: Tensor, b: Tensor, inputMean: Tensor, inputVar: Tensor): Tensor {
    return recordOp('BatchNormalization', [x, scale, b, inputMean, inputVar], {
      epsilon: this.epsilon,
      momentum: this.momentum,
    });
  }
}

/**
 * Inherits from BaseNorm. Parametrized by normalized_shape.
 */
export class LayerNormalization extends BaseNorm {
  public normalizedShape: number[];
  public axis: number;

  constructor(normalizedShape: number[], epsilon: number = 1e-5) {
    super(epsilon);
    this.normalizedShape = normalizedShape;
    this.axis = -normalizedShape.length;
  }

  /**
   * Applies Layer Normalization.
   */
  call(x: Tensor, scale: Tensor, b?: Tensor): Tensor {
    const inputs = b ? [x, scale, b] : [x, scale];
    return recordOp('LayerNormalization', inputs, {
      axis: this.axis,
      epsilon: this.epsilon,
    });
  }
}

/**
 * Inherits from BaseNorm. Standardizes LLaMA/Gemma variance-only scaling.
 */
export class RMSNorm extends BaseNorm {
  public normalizedShape: number[];

  constructor(normalizedShape: number[], epsilon: number = 1e-5) {
    super(epsilon);
    this.normalizedShape = normalizedShape;
  }

  /**
   * Applies RMS Normalization.
   */
  call(x: Tensor, scale: Tensor): Tensor {
    return recordOp('RMSNormalization', [x, scale], {});
  }
}

/**
 * Maps via generalized Reshape -> LayerNorm -> Reshape subgraphs.
 */
export class GroupNorm extends BaseNorm {
  public numGroups: number;
  public numChannels: number;

  constructor(numGroups: number, numChannels: number, epsilon: number = 1e-5) {
    super(epsilon);
    this.numGroups = numGroups;
    this.numChannels = numChannels;
  }

  /**
   * Applies Group Normalization.
   */
  call(x: Tensor, scale: Tensor, b: Tensor): Tensor {
    return recordOp('GroupNormalization', [x, scale, b], {
      epsilon: this.epsilon,
      num_groups: this.numGroups,
    });
  }
}

/**
 * Instance Normalization.
 */
export class InstanceNorm extends BaseNorm {
  public numFeatures: number;

  constructor(numFeatures: number, epsilon: number = 1e-5) {
    super(epsilon);
    this.numFeatures = numFeatures;
  }

  /**
   * Applies Instance Normalization.
   */
  call(x: Tensor, scale: Tensor, b: Tensor): Tensor {
    return recordOp('InstanceNormalization', [x, scale, b], {
      epsilon: this.epsilon,
    });
  }
}

/**
 * Abstract base class for element-wise non-linearities.
 */
export abstract class BaseActivation {
  abstract call(x: Tensor): Tensor;

  generateLUT(numPoints: number = 256, rangeMin: number = -8.0, rangeMax: number = 8.0): Tensor {
    return recordOp('Constant', [], {
      lut_range: [rangeMin, rangeMax],
      lut_points: numPoints,
      activation: this.constructor.name,
    });
  }
}

export class Relu extends BaseActivation {
  call(x: Tensor): Tensor {
    return recordOp('Relu', [x]);
  }
}

export class Sigmoid extends BaseActivation {
  call(x: Tensor): Tensor {
    return recordOp('Sigmoid', [x]);
  }
}

export class Tanh extends BaseActivation {
  call(x: Tensor): Tensor {
    return recordOp('Tanh', [x]);
  }
}

export class LeakyRelu extends BaseActivation {
  public alpha: number;
  constructor(alpha: number = 0.01) {
    super();
    this.alpha = alpha;
  }
  call(x: Tensor): Tensor {
    return recordOp('LeakyRelu', [x], { alpha: this.alpha });
  }
}

export class Gelu extends BaseActivation {
  public approximate: string;
  constructor(approximate: string = 'none') {
    super();
    this.approximate = approximate;
  }
  call(x: Tensor): Tensor {
    return recordOp('Gelu', [x], { approximate: this.approximate });
  }
}

export class Silu extends BaseActivation {
  call(x: Tensor): Tensor {
    return recordOp('Swish', [x]);
  }
}

export class Swish extends Silu {}

export class Mish extends BaseActivation {
  call(x: Tensor): Tensor {
    return recordOp('Mish', [x]);
  }
}

export abstract class ConvFamily {
  public inChannels: number;
  public outChannels: number;
  public kernelSize: number | number[];
  public stride: number | number[];
  public padding: number | number[];
  public dilation: number | number[];
  public groups: number;
  public bias: boolean;

  constructor(
    inChannels: number,
    outChannels: number,
    kernelSize: number | number[],
    stride: number | number[] = 1,
    padding: number | number[] = 0,
    dilation: number | number[] = 1,
    groups: number = 1,
    bias: boolean = true,
  ) {
    this.inChannels = inChannels;
    this.outChannels = outChannels;
    this.kernelSize = kernelSize;
    this.stride = stride;
    this.padding = padding;
    this.dilation = dilation;
    this.groups = groups;
    this.bias = bias;
  }

  abstract call(x: Tensor, w: Tensor, b?: Tensor): Tensor;
}

export class ConvND extends ConvFamily {
  public dims: number;
  constructor(
    dims: number,
    inChannels: number,
    outChannels: number,
    kernelSize: number | number[],
    stride: number | number[] = 1,
    padding: number | number[] = 0,
    dilation: number | number[] = 1,
    groups: number = 1,
    bias: boolean = true,
  ) {
    super(inChannels, outChannels, kernelSize, stride, padding, dilation, groups, bias);
    this.dims = dims;
  }

  override call(x: Tensor, w: Tensor, b?: Tensor): Tensor {
    const ks = Array.isArray(this.kernelSize)
      ? this.kernelSize
      : Array(this.dims).fill(this.kernelSize);
    const st = Array.isArray(this.stride) ? this.stride : Array(this.dims).fill(this.stride);
    const pa = Array.isArray(this.padding) ? this.padding : Array(this.dims * 2).fill(this.padding);
    const di = Array.isArray(this.dilation) ? this.dilation : Array(this.dims).fill(this.dilation);

    const inputs = b ? [x, w, b] : [x, w];
    return recordOp('Conv', inputs, {
      kernel_shape: ks,
      strides: st,
      pads: pa,
      dilations: di,
      group: this.groups,
    });
  }
}

export class DepthwiseConv extends ConvND {
  constructor(
    dims: number,
    channels: number,
    kernelSize: number | number[],
    stride: number | number[] = 1,
    padding: number | number[] = 0,
    dilation: number | number[] = 1,
    bias: boolean = true,
  ) {
    super(dims, channels, channels, kernelSize, stride, padding, dilation, channels, bias);
  }
}

export class MatMul {
  call(x: Tensor, y: Tensor): Tensor {
    return recordOp('MatMul', [x, y]);
  }
}

export class Gemm {
  public alpha: number;
  public beta: number;
  public transA: number;
  public transB: number;

  constructor(alpha: number = 1.0, beta: number = 1.0, transA: number = 0, transB: number = 0) {
    this.alpha = alpha;
    this.beta = beta;
    this.transA = transA;
    this.transB = transB;
  }

  call(x: Tensor, y: Tensor, c?: Tensor): Tensor {
    const inputs = c ? [x, y, c] : [x, y];
    return recordOp('Gemm', inputs, {
      alpha: this.alpha,
      beta: this.beta,
      transA: this.transA,
      transB: this.transB,
    });
  }
}

export class MultiHeadAttention {
  public numHeads: number;
  public qkvBias: boolean;
  public outBias: boolean;

  constructor(numHeads: number, qkvBias: boolean = true, outBias: boolean = true) {
    this.numHeads = numHeads;
    this.qkvBias = qkvBias;
    this.outBias = outBias;
  }

  call(q: Tensor, k: Tensor, v: Tensor, mask?: Tensor): Tensor {
    const inputs = mask ? [q, k, v, mask] : [q, k, v];
    return recordOp('Attention', inputs, { num_heads: this.numHeads });
  }
}

export class FlashAttention extends MultiHeadAttention {
  override call(q: Tensor, k: Tensor, v: Tensor, mask?: Tensor): Tensor {
    const inputs = mask ? [q, k, v, mask] : [q, k, v];
    return recordOp('FlashAttention', inputs, { num_heads: this.numHeads });
  }
}

export class GroupedQueryAttention extends MultiHeadAttention {
  public numKvHeads: number;

  constructor(
    numHeads: number,
    numKvHeads: number,
    qkvBias: boolean = false,
    outBias: boolean = false,
  ) {
    super(numHeads, qkvBias, outBias);
    this.numKvHeads = numKvHeads;
  }

  override call(q: Tensor, k: Tensor, v: Tensor, mask?: Tensor): Tensor {
    const inputs = mask ? [q, k, v, mask] : [q, k, v];
    return recordOp('GroupedQueryAttention', inputs, {
      num_heads: this.numHeads,
      num_kv_heads: this.numKvHeads,
    });
  }
}

export class RoPE {
  public dim: number;
  public base: number;
  public maxSeqLen: number;

  constructor(dim: number, base: number = 10000.0, maxSeqLen: number = 2048) {
    this.dim = dim;
    this.base = base;
    this.maxSeqLen = maxSeqLen;
  }

  call(x: Tensor, pos: Tensor): Tensor {
    return recordOp('RoPE', [x, pos], { dim: this.dim, base: this.base });
  }
}

export class AlibiBias {
  public numHeads: number;

  constructor(numHeads: number) {
    this.numHeads = numHeads;
  }

  call(mask: Tensor): Tensor {
    return recordOp('AlibiBias', [mask], { num_heads: this.numHeads });
  }
}
